File: mcp-server/internal/domains/search/tools.go
// Package search provides the "search" MCP tool backed by stigmer-server's
// SearchService.search RPC.
//
// This single tool covers listing, full-text search, and cross-kind discovery
// — eliminating the need for separate list_agents / list_skills / list_workflows
// tools. The caller controls behavior through the combination of parameters:
//
//   - List agents in org:   {kinds: ["agent"], org: "acme"}
//   - Search across all:    {query: "kubernetes"}
//   - Filtered search:      {kinds: ["agent","skill"], query: "security", org: "acme"}
//
// Results are returned as a JSON object matching the SearchResponse protobuf
// message, serialized via protojson for clean field names and well-known-type
// handling (e.g. google.protobuf.Timestamp → RFC 3339 strings). Each result
// entry is enriched with a resource_uri field that MCP clients can pass
// directly to resources/read, bridging the discovery-to-read workflow.
package search

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
	"github.com/stigmer/stigmer/mcp-server/internal/domains"

	"github.com/stigmer/stigmer/mcp-server/internal/auth"
	stigmergrpc "github.com/stigmer/stigmer/mcp-server/internal/grpc"

	apiresourcekind "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	rpcpb "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/rpc"
	searchv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/search/v1"
)

// SearchInput defines the parameters that an MCP client can pass to the
// "search" tool.  Field tags drive the JSON Schema that the MCP SDK
// auto-generates and sends to clients during tool discovery.
type SearchInput struct {
	// Kinds restricts results to specific resource types.
	// Valid values: "agent", "skill", "mcp_server", "workflow".
	// Empty means search across all searchable kinds (discover mode).
	Kinds []string `json:"kinds,omitempty" jsonschema:"description=Resource kinds to search. Valid: agent, skill, mcp_server, workflow. Empty searches all."`

	// Query is a free-text search string. Empty returns all accessible
	// resources (list mode).
	Query string `json:"query,omitempty" jsonschema:"description=Full-text search query. Empty lists all accessible resources."`

	// Org scopes results to a single organization. Empty searches across
	// all organizations the caller has access to.
	Org string `json:"org,omitempty" jsonschema:"description=Organization slug to scope the search. Empty searches all accessible orgs."`

	// ExcludePublic, when true, hides platform-provided public resources.
	ExcludePublic bool `json:"exclude_public,omitempty" jsonschema:"description=Exclude public/platform resources from results."`

	// PageSize controls how many results per page (default 20, max 100).
	PageSize int32 `json:"page_size,omitempty" jsonschema:"description=Results per page (default 20, max 100)."`

	// PageNum selects which page to return (1-indexed, default 1).
	PageNum int32 `json:"page_num,omitempty" jsonschema:"description=Page number (1-indexed, default 1)."`
}

// Tool returns the MCP tool definition for registration with the server.
func Tool() *mcp.Tool {
	return &mcp.Tool{
		Name: "search",
		Description: "Search and list Stigmer resources (agents, skills, MCP servers, workflows). " +
			"Set 'kinds' to filter by resource type. Set 'query' for full-text search. " +
			"Set 'org' to scope to an organization. Omit 'query' to list all accessible resources.",
	}
}

// Handler implements the "search" tool.  The serverAddress is captured at
// registration time; the API key is read from the context at call time.
func Handler(serverAddress string) func(context.Context, *mcp.CallToolRequest, *SearchInput) (*mcp.CallToolResult, any, error) {
	return func(ctx context.Context, req *mcp.CallToolRequest, input *SearchInput) (*mcp.CallToolResult, any, error) {
		conn, err := stigmergrpc.NewConnection(serverAddress, auth.APIKey(ctx))
		if err != nil {
			return nil, nil, fmt.Errorf("search: %w", err)
		}
		defer conn.Close()

		// Map user-supplied kind strings to the proto enum values.
		kinds, err := parseKinds(input.Kinds)
		if err != nil {
			return nil, nil, err
		}

		grpcReq := &searchv1.SearchRequest{
			Kinds:         kinds,
			Query:         input.Query,
			Org:           input.Org,
			ExcludePublic: input.ExcludePublic,
		}
		if input.PageSize > 0 || input.PageNum > 0 {
			grpcReq.Page = &rpcpb.PageInfo{
				Size: input.PageSize,
				Num:  input.PageNum,
			}
		}

		rpcCtx, cancel := context.WithTimeout(ctx, stigmergrpc.DefaultRPCTimeout)
		defer cancel()

		client := searchv1.NewSearchServiceClient(conn)
		resp, err := client.Search(rpcCtx, grpcReq)
		if err != nil {
			desc := "search results"
			if input.Org != "" {
				desc = fmt.Sprintf("search results in org %q", input.Org)
			}
			return nil, nil, domains.RPCError(err, desc)
		}

		text, err := enrichSearchResponse(resp)
		if err != nil {
			return nil, nil, err
		}
		return &mcp.CallToolResult{
			Content: []mcp.Content{&mcp.TextContent{Text: text}},
		}, nil, nil
	}
}

// enrichSearchResponse serializes the search response to JSON and injects a
// resource_uri field into each entry that has a registered MCP resource
// template. Entries whose kind has no template (e.g. mcp_server) are left
// without a resource_uri.
//
// When the response has no entries, it short-circuits to domains.MarshalJSON
// to avoid unnecessary round-tripping.
func enrichSearchResponse(resp *searchv1.SearchResponse) (string, error) {
	if len(resp.GetEntries()) == 0 {
		return domains.MarshalJSON(resp)
	}

	raw, err := domains.MarshalOptions.Marshal(resp)
	if err != nil {
		return "", fmt.Errorf("protojson marshal: %w", err)
	}

	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return "", fmt.Errorf("json unmarshal: %w", err)
	}

	entries, _ := data["entries"].([]any)
	for i, e := range entries {
		entry, ok := e.(map[string]any)
		if !ok {
			continue
		}
		if i >= len(resp.Entries) {
			break
		}
		r := resp.Entries[i]
		uri := domains.BuildResourceURI(r.Kind.String(), r.Org, r.Slug)
		if uri != "" {
			entry["resource_uri"] = uri
		}
	}

	out, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", fmt.Errorf("json marshal: %w", err)
	}
	return string(out), nil
}

// knownKinds maps user-friendly kind names to proto enum values.
var knownKinds = map[string]apiresourcekind.ApiResourceKind{
	"agent":      apiresourcekind.ApiResourceKind_agent,
	"skill":      apiresourcekind.ApiResourceKind_skill,
	"mcp_server": apiresourcekind.ApiResourceKind_mcp_server,
	"workflow":   apiresourcekind.ApiResourceKind_workflow,
}

// parseKinds converts user-supplied kind strings to proto enum values.
func parseKinds(raw []string) ([]apiresourcekind.ApiResourceKind, error) {
	if len(raw) == 0 {
		return nil, nil
	}
	out := make([]apiresourcekind.ApiResourceKind, 0, len(raw))
	for _, s := range raw {
		k, ok := knownKinds[s]
		if !ok {
			return nil, fmt.Errorf("unknown resource kind %q; valid kinds: agent, skill, mcp_server, workflow", s)
		}
		out = append(out, k)
	}
	return out, nil
}

File: mcp-server/src/domains/search/tools.ts
// The "search" MCP tool, backed by SearchService.search.
// Go parity: mcp-server/internal/domains/search/tools.go.
//
// One tool covers listing, full-text search, and cross-kind discovery (replacing
// per-kind list_* tools); the combination of parameters selects the behavior.
// Each result entry is enriched with a resource_uri that clients can hand
// straight to resources/read, bridging discovery to read.

import { toJson, type MessageInitShape } from "@bufbuild/protobuf";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";
import {
  type SearchRequestSchema,
  type SearchResponse,
  SearchResponseSchema,
} from "@stigmer/protos/ai/stigmer/search/v1/io_pb";
import { SearchService } from "@stigmer/protos/ai/stigmer/search/v1/query_pb";
import { z } from "zod";

import { resolveToken, withClient, type BackendTarget } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { buildResourceURI } from "../resourceuri.js";
import { rpcError } from "../rpcerr.js";
import { textOrError } from "../toolresult.js";

/**
 * The searchable kinds and their proto enum values. Deliberately a curated
 * subset (not every ApiResourceKind): the kinds an MCP client can also read
 * and manage through tools. The backend's extractor registry supports more
 * (sessions, executions, ...) — add here only alongside a tool surface.
 */
const knownKinds: Readonly<Record<string, ApiResourceKind>> = {
  agent: ApiResourceKind.agent,
  environment: ApiResourceKind.environment,
  skill: ApiResourceKind.skill,
  mcp_server: ApiResourceKind.mcp_server,
  workflow: ApiResourceKind.workflow,
};

interface SearchArgs {
  readonly kinds?: string[];
  readonly query?: string;
  readonly org?: string;
  readonly excludePublic?: boolean;
  readonly pageSize?: number;
  readonly pageNum?: number;
}

/** Register the search tool; returns the registered tool names. */
export function registerSearchTools(server: McpServer, target: BackendTarget): string[] {
  server.registerTool(
    "search",
    {
      description:
        "Search and list Stigmer resources (agents, skills, MCP servers, workflows, " +
        "environments). Set 'kinds' to filter by resource type. Set 'query' for full-text search. " +
        "Set 'org' to scope to an organization. Omit 'query' to list all accessible resources.",
      inputSchema: {
        kinds: z
          .array(z.string())
          .optional()
          .describe(
            "Resource kinds to search. Valid: agent, skill, mcp_server, workflow, environment. Empty searches all.",
          ),
        query: z
          .string()
          .optional()
          .describe("Full-text search query. Empty lists all accessible resources."),
        org: z
          .string()
          .optional()
          .describe("Organization slug to scope the search. Empty searches all accessible orgs."),
        exclude_public: z
          .boolean()
          .optional()
          .describe("Exclude public/platform resources from results."),
        page_size: z.number().int().optional().describe("Results per page (default 20, max 100)."),
        page_num: z.number().int().optional().describe("Page number (1-indexed, default 1)."),
      },
    },
    (args, extra) =>
      textOrError(() =>
        search(target.serverAddress, resolveToken(extra, target.apiKey), {
          kinds: args.kinds,
          query: args.query,
          org: args.org,
          excludePublic: args.exclude_public,
          pageSize: args.page_size,
          pageNum: args.page_num,
        }),
      ),
  );

  return ["search"];
}

/** Run the search RPC and return the enriched protojson result. */
async function search(serverAddress: string, token: string, args: SearchArgs): Promise<string> {
  const kinds = parseKinds(args.kinds);
  return withClient(SearchService, serverAddress, token, async (client, callOptions) => {
    const req: MessageInitShape<typeof SearchRequestSchema> = {
      kinds,
      query: args.query ?? "",
      org: args.org ?? "",
      excludePublic: args.excludePublic ?? false,
    };
    // Attach pagination only when explicitly requested, letting the server apply
    // its own defaults otherwise (Go does the same).
    if ((args.pageSize ?? 0) > 0 || (args.pageNum ?? 0) > 0) {
      req.page = { size: args.pageSize ?? 0, num: args.pageNum ?? 0 };
    }

    let resp: SearchResponse;
    try {
      resp = await client.search(req, callOptions);
    } catch (err) {
      throw rpcError(err, args.org ? `search results in org "${args.org}"` : "search results");
    }
    return enrichSearchResponse(resp);
  });
}

/** Convert user-supplied kind strings to proto enum values, rejecting unknowns. */
function parseKinds(raw: string[] | undefined): ApiResourceKind[] {
  if (!raw || raw.length === 0) {
    return [];
  }
  return raw.map((s) => {
    const kind = knownKinds[s];
    if (kind === undefined) {
      throw new Error(
        `unknown resource kind "${s}"; valid kinds: agent, skill, mcp_server, workflow, environment`,
      );
    }
    return kind;
  });
}

/**
 * Serialize the response and inject a resource_uri into each entry whose kind
 * has a registered resource template. Empty responses short-circuit to the plain
 * marshal to avoid the walk.
 */
function enrichSearchResponse(resp: SearchResponse): string {
  if (resp.entries.length === 0) {
    return toProtoJson(SearchResponseSchema, resp);
  }

  const data = toJson(SearchResponseSchema, resp, { useProtoFieldName: true }) as Record<
    string,
    unknown
  >;
  const entries = Array.isArray(data.entries)
    ? (data.entries as Array<Record<string, unknown>>)
    : [];
  for (let i = 0; i < entries.length && i < resp.entries.length; i++) {
    const r = resp.entries[i]!;
    const uri = buildResourceURI(ApiResourceKind[r.kind] ?? "", r.org, r.slug);
    if (uri !== "") {
      entries[i]!.resource_uri = uri;
    }
  }
  return JSON.stringify(data, null, 2);
}
